Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp

```diff
@@ -185,8 +185,25 @@ namespace AZ
             return;
         }
 
+        // Get the hash before locking since another thread could be deleting the object within the lock
+        Internal::NameData::Hash hash = nameData->GetHash();
```

Comment:
Something still doesn't seem right here.

`nameData` can still be deleted before even getting to this line

Therefore I think the right solution is to retrieve the hash value, before decrementing the NameData count in the [NameData::release()](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Name/Internal/NameData.cpp#L40)

Here is the scenario
- T1, gets into TryReleaseName, but has not looked up the hash yet
- T2 gets into MakeName, acquires the lock, returns a new Name that increments the counter
- T2 Name decrements the counter, gets into TryReleaseName
- T2 looks up the hash value
- T2 gets the acquires the lock
- T2 Deletes the NameData
- T1 **now looks up the hash** and accesses invalid memory due to nameData being deleted

Reply:
you are right, let me edit it